File: programmers/javascript/level2/다리를지나는트럭.js
const val1 = 100;
const val2 = 100;
const val3 = [10, 10, 10, 10, 10, 10, 10, 10, 10, 10];

const sumValue = (arr) => {
  let sum = 0;
  for (let i = 0; i < arr.length; i += 1) {
    sum += arr[i][0];
  }
  return sum;
};

const solution = (bridgeLength, weight, truckWeights) => {
  let answer = 0;
  let temp = [];
  while (truckWeights.length !== 0 || temp.length !== 0) {
    answer += 1;
    if (
      temp.length === 0 ||
      (sumValue(temp) + truckWeights[0] <= weight &&
        temp.length <= bridgeLength)
    ) {
      temp.unshift([truckWeights[0], 0]);
      truckWeights.shift();
    }
    for (let i = 0; i < temp.length; i += 1) {
      temp[i][1] += 1;
    }
    temp = temp.filter((it) => it[1] < bridgeLength);
  }
  answer += 1;
  return answer;
};

console.log(solution(val1, val2, val3));

File: python/programmers/level2/다리를지나는트럭.py
from collections import deque


def solution(bridge_length, weight, truck_weights):
    leftTruck = deque(truck_weights)
    answer = 0
    q = deque()
    nowTruckWeight = 0
    # 아직 다리를 건널 트럭이 있을때까지 루프
    while leftTruck or q:
        # 트럭이 다리를 다 건넜으면 빼준다.
        if q and q[0][1] == bridge_length:
            nowTruckWeight -= q[0][0]
            q.popleft()
        # 건너지 못한 트럭이있고 다리에 오를수있으면 트럭을 다리에 올린다.
        if leftTruck and nowTruckWeight + leftTruck[0] <= weight and len(q) < bridge_length:
            x = leftTruck.popleft()
            q.append([x, 0])
            nowTruckWeight += x

        for x in range(len(q)):
            q[x][1] += 1
        # print("%d초" % answer)
        # print(q)
        answer += 1

    return answer


print(solution(100, 100, [10, 10, 10, 10, 10, 10, 10, 10, 10, 10]))
